python-test-samples/async-architectures/async-lambda-dynamodb/README.md

```diff
@@ -0,0 +1,70 @@
+
+![AWS Lambda and AmazonDynamoDB](../img/lambda-dynamo.png)
+
+[[top]](#asynchronous-integration-test-with-lambda-event-listener-and-dynamodb)
+
+## Review the System Under Test
+
+The System Under Test (SUT) in this example is an asynchronous text processor. It contains a source S3 bucket that receives a text file. A Lambda function is configured to be notified when the putObject event is executed on this bucket. The Lambda function gets the file from the source bucket, transforms the contents of the file to uppercase, then puts the file into a destination S3 bucket.
+
+![S3 to Lambda to S3](../img/s3-lambda-s3.png)
+
+Your goal is to test this asynchronous process. You will use the Lambda Event Listener and DynamoDB test pattern to test the process. You will deploy the following resources:
+
+* S3 Source Bucket
+* Lambda function text processor
+* S3 Destination Bucket
+* Lambda event listener (test environments only)
+* DynamoDB results storage (test environments only)
+
+[[top]](#asynchronous-integration-test-with-lambda-event-listener-and-dynamodb)
+
+## Prerequisites
+The SAM CLI is an extension of the AWS CLI that adds functionality for building and testing serverless applications. It contains features for building your appcation locally, deploying it to AWS, and emulating AWS services locally to support automated unit tests.  
+
+To use the SAM CLI, you need the following tools.
+
+- SAM CLI - [Install the SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html)
+- Python 3 - [Install Python 3](https://www.python.org/downloads/)
+- Docker - [Install Docker community edition](https://hub.docker.com/search/?type=edition&offering=community)
+
+This project uses the `make` build automation tool to provide convenient commands for building, deploying and testing this sample application. To view the available commands make sure this utility is installed and run the `make` command in the terminal from the same directory as the `Makefile`.
+
+- Make - [Read the documentation](https://www.gnu.org/software/make/manual/html_node/index.html)
+
+[[top]](#asynchronous-integration-test-with-lambda-event-listener-and-dynamodb)
+
+## Build the project
+
+- `make build` - Build the project locally.
+
+[[top]](#asynchronous-integration-test-with-lambda-event-listener-and-dynamodb)
+
+## Deploy project resources to the cloud
+
+- `make deploy.g` -  Deploy the project to the AWS cloud for the first time.
+- `make deploy`   -  Deploy subsequent changes to the AWS cloud. This command uses a saved configuration file for configurations.
+
+[[top]](#asynchronous-integration-test-with-lambda-event-listener-and-dynamodb)
+
+## Run the tests
+Install a virtual environment to install the development dependencies and execute your tests.
+Run these four commands in sequence.   
+
+- `make install` - Install Pipenv
+- `make shell`   - Spawn a virtual environment shell
```

Comment:
Using the commands as a "new" programmer would, make shell receives this error:

[*] Type the 'deactivate' command to exit this shell
Shell for UNKNOWN_VIRTUAL_ENVIRONMENT already activated.
No action taken to avoid nested environments.
make[1]: *** [shell] Error 1

Reply:
I notice that the pyenv shell gets into a confused state if it's not closed properly. The `exit` command seems to work much better than `deactivate` despite the copy in the warning. Can you try that command and see if it works? I have added this instruction to the README. https://github.com/aws-samples/serverless-test-samples/commit/4bc04fb1cd7663dc4c94617ee3a7f19acd1a9768